feat: add training task for qna (#3740)
recognai/rubrix: docs/_source/guides/llms/practical_guides/fine_tune.md

Underneath you can find a sample of an extractive QnA dataset underneath:

```batch
{
    'question': 'To whom did the Virgin Mary allegedly appear in 1858 in Lourdes France?',
    'context': 'Architecturally, the school has a Catholic character. Atop the Main Building\'s gold dome is a golden statue of the Virgin Mary. Immediately in front of the Main Building and facing it, is a copper statue of Christ with arms upraised with the legend "Venite Ad Me Omnes". Next to the Main Building is the Basilica of the Sacred Heart. Immediately behind the basilica is the Grotto, a Marian place of prayer and reflection. It is a replica of the grotto at Lourdes, France where the Virgin Mary reputedly appeared to Saint Bernadette Soubirous in 1858. At the end of the main drive (and in a direct line that connects through 3 statues and the Gold Dome), is a simple, modern stone statue of Mary.',
    'answers': 'Saint Bernadette Soubirous',
}
```

```{note}
Officially, answers need to be passed as a list of `{'answer_start': int, 'text': str}`-dicts. However, we only support a string, where the `answer_start` is inferred from the `context` and `text`-field.
```

> We then use either `question-context-answet`-set or a `formatting_func` to further fine-tune the model.

Reviewer: ```suggestion
We then use either `question-context-answer`-set or a `formatting_func` to further fine-tune the model.
```

#### Training

**Data Preparation**

```python
import argilla as rg
from datasets import Dataset

feedback_dataset = rg.FeedbackDataset.from_huggingface("argilla/squad")
```

> We can use a default configuraiton where we initialize the `TrainingTask.for_question_answering` using the `question-context-answer`-set from the dataset. We also offer the option to provide a `formatting_func` to the `TrainingTask.for_question_asnwering`. This function is applied to each sample in the dataset and can be used for advanced preprocessing and data formatting. The function should return a `question-context-answer`-set as `str-str-str`.

Reviewer: ```suggestion
We can use a default configuration where we initialize the `TrainingTask.for_question_answering` using the `question-context-answer`-set from the dataset. We also offer the option to provide a `formatting_func` to the `TrainingTask.for_question_asnwering`. This function is applied to each sample in the dataset and can be used for advanced preprocessing and data formatting. The function should return a `question-context-answer`-set as `str-str-str`.
```

I also think that the "-set" phrasing that you use throughout these docs is very confusing, it makes me think of a Python set. Perhaps "pair", "tuple", or "triple" is clearer.

:::: {tab-set}

::: {tab-item} question-context-answer-set

```python
from argilla.feedback import TrainingTask

task = TrainingTask.for_question_answering(

>     question=feedback_dataset.question_by_name("question"),
>     context=feedback_dataset.question_by_name("context"),

Reviewer: ```suggestion
    question=feedback_dataset.field_by_name("question"),
    context=feedback_dataset.field_by_name("context"),
```

    answer=feedback_dataset.question_by_name("answer"),
)
```

:::

::: {tab-item} formatting_func

```python
from argilla.feedback import TrainingTask

> def formatting_func(sample):
>     responses = []
>     question = sample["question"]
>     context = sample["context"]
>     for answer in sample["answer"]:
>         if not all([question, context, answer["value"]]):
>             continue
>         responses.append((question, context, answer["value"]))
>     return responses

Reviewer: ```suggestion
def formatting_func(sample):
    question = sample["question"]
    context = sample["context"]
    for answer in sample["answer"]:
        if not all([question, context, answer["value"]]):
            continue
        yield question, context, answer["value"]
```
I'm not sure if this is supported in the actual implementation - but it should be.

    if sample["response"]:
        chat = str(uuid4())
        user_message = user_message_prompt.format(context_str=sample["context"], query_str=sample["user-message"])
        yield [
            (chat, "0", "system", system_prompt),
            (chat, "1", "user", user_message),
            (chat, "2", "assistant", sample["response"][0]["value"])
        ]
    else:

>         yield None

Reviewer: This `else` branch should just be removed
